src/common/statistics/StatisticsManager.ts

```diff
@@ -0,0 +1,394 @@
+  RelayRegisteredEventInfo,
+  RelayServerRegistered,
+  StakeAdded,
+  StakeAddedEventInfo,
+  StakeChangeEvent,
+  StakePenalized,
+  StakePenalizedEventInfo,
+  StakeUnlocked,
+  StakeUnlockedEventInfo,
+  StakeWithdrawn,
+  StakeWithdrawnEventInfo,
+  TransactionRejectedByPaymaster,
+  TransactionRejectedByPaymasterEventInfo,
+  TransactionRelayed,
+  TransactionRelayedEventInfo,
+  allStakeManagerEvents
+} from '../types/GSNContractsDataTypes'
+
+import {
+  EventTransactionInfo,
+  GSNStatistics,
+  PaymasterInfo,
+  PingResult,
+  RecipientInfo,
+  RelayHubConstructorParams,
+  RelayHubEvents,
+  RelayServerInfo,
+  RelayServerRegistrationInfo,
+  RelayServerStakeStatus,
+  SenderInfo,
+  StakeMangerEvents
+} from '../types/GSNStatistics'
+
+import { gsnRuntimeVersion } from '../Version'
+import { LoggerInterface } from '../LoggerInterface'
+
+export default class StatisticsManager {
+  private readonly contractInteractor: ContractInteractor
+  private readonly httpClient: HttpClient
+  private readonly logger: LoggerInterface
+
+  private allStakeManagerEvents!: StakeMangerEvents
+  private allRelayHubEvents!: RelayHubEvents
+
+  constructor (
+    contractInteractor: ContractInteractor,
+    httpClient: HttpClient,
+    logger: LoggerInterface) {
+    this.contractInteractor = contractInteractor
+    this.httpClient = httpClient
+    this.logger = logger
+  }
+
+  async gatherStatistics (): Promise<GSNStatistics> {
+    const chainId = this.contractInteractor.chainId
+    const blockNumber = await this.contractInteractor.getBlockNumber()
+    const relayHubConstructorParams = await this.getRelayHubConstructorParams()
+
+    const stakeManagerAddress = this.contractInteractor.stakeManagerAddress()
+    const totalStakesByRelays = await this.contractInteractor.getBalance(stakeManagerAddress)
+    await this.fetchStakeManagerEvents()
+    await this.fetchRelayHubEvents()
+
+    const relayServers = await this.getRelayServersInfo()
+    const paymasters = await this.getPaymastersInfo()
+    const senders = await this.getSendersInfo()
+    const recipients = await this.getRecipientsInfo()
+    const totalGasPaidViaGSN = '0'
+
+    const runtimeVersion = gsnRuntimeVersion
+    const contractsDeployment = this.contractInteractor.getDeployment()
+    const deploymentVersions = await this.contractInteractor.resolveDeploymentVersions()
+    const deploymentBalances = await this.contractInteractor.queryDeploymentBalances()
+    return {
+      relayHubEvents: this.allRelayHubEvents,
+      stakeManagerEvents: this.allStakeManagerEvents,
+      relayHubConstructorParams,
+      deploymentBalances,
+      chainId,
+      runtimeVersion,
+      deploymentVersions,
+      contractsDeployment,
+      blockNumber,
+      totalStakesByRelays,
+      paymasters,
+      senders,
+      recipients,
+      relayServers,
+      totalGasPaidViaGSN
+    }
+  }
+
+  async fetchRelayHubEvents (): Promise<void> {
+    const transactionDepositedEventsData =
```

Comment:
I lost you here: you have a nice `extractTransactionInfos()` to filter out specific type from "all events" - but still, you do getPastEventsForHub for each event separately...
do gpefh once, and filter its output 5 times
(ok, if you really want, then split the TransactionRelayed out  - but all the rest can be slurped together)